order-book/src/main/java/com/market/banica/order/book/model/ItemMarket.java

```diff
@@ -37,22 +54,116 @@ public void removeUntrackedItem(String itemName) {
         allItems.remove(itemName);
     }
 
+    public void updateItem(Aurora.AuroraResponse response) {
+        if (response.getMessage().is(TickResponse.class)) {
+            TickResponse tickResponse;
+
+            try {
+                tickResponse = response.getMessage().unpack(TickResponse.class);
+            } catch (InvalidProtocolBufferException e) {
+                throw new IncorrectResponseException("Response is not correct!");
+            }
+
+            Item item = new Item();
+            item.setPrice(tickResponse.getPrice());
+            item.setQuantity(tickResponse.getQuantity());
+            item.setOrigin(tickResponse.getOrigin());
+
+            productsQuantity.merge(tickResponse.getGoodName(), tickResponse.getQuantity(), Long::sum);
+
+            Set<Item> itemSet = allItems.get(tickResponse.getGoodName());
+            if (itemSet != null) {
+                if (itemSet.contains(item)) {
+                    Item presentItem = itemSet.stream().filter(item1 -> Double.compare(item1.getPrice(), item.getPrice()) == 0
+                            && item1.getOrigin().equals(item.getOrigin())).findFirst().get();
+                    presentItem.setQuantity(presentItem.getQuantity() + item.getQuantity());
+                    return;
+                }
+
+                itemSet.add(item);
+            } else {
+                LOGGER.error("Item: {} is not being tracked and cannot be added to itemMarket!",
+                        tickResponse.getGoodName());
+            }
+
+            LOGGER.info("Products data updated!");
+        } else {
+            throw new IncorrectResponseException("Response is not correct!");
+        }
+    }
+
+    public List<OrderBookLayer> getRequestedItem(String itemName, long quantity) {
+        LOGGER.info("Getting requested item: {} with quantity: {}", itemName, quantity);
+        TreeSet<Item> items = this.allItems.get(itemName);
+
+        if (items == null || productsQuantity.get(itemName) < quantity) {
+            return Collections.emptyList();
+        }
+
+        List<OrderBookLayer> layers;
+        try {
+            lock.writeLock().lock();
+
+            layers = new ArrayList<>();
+
+            Iterator<Item> iterator = items.iterator();
+            long itemLeft = quantity;
+
+            while (iterator.hasNext() && itemLeft > 0) {
+                Item currentItem = iterator.next();
+
+                OrderBookLayer.Builder currentLayer = OrderBookLayer.newBuilder()
```

Comment:
all this logic from like 115 to line 130 is not directly related to getRequestedItem, it could be in a separate method

Reply:
Done